apps/sms/services/js/service_manager.js

```diff
@@ -0,0 +1,239 @@
+    ['moz-mobile-connections', MozMobileConnectionsServiceInitializer],
+    ['moz-settings', MozSettingsServiceInitializer]
+  ]);
+
+  function initializeShims(endpoint, upgradeKey) {
+    if (endpoint.MozMobileMessageShim) {
+      endpoint.MozMobileMessageShim.init(
+        endpoint.navigator.mozMobileMessage,
+        upgradeKey ? new endpoint.BroadcastChannel(
+          `${endpoint.MozMobileMessageShim.name}-${upgradeKey}`
+        ) : endpoint
+      );
+
+      debug('Initialized MozMobileMessageShim (upgrade key: %s)', upgradeKey);
+    }
+
+    if (endpoint.MozSettingsShim) {
+      endpoint.MozSettingsShim.init(
+        endpoint.navigator.mozSettings,
+        upgradeKey ? new endpoint.BroadcastChannel(
+          `${endpoint.MozSettingsShim.name}-${upgradeKey}`
+        ) : endpoint
+      );
+
+      debug('Initialized MozSettingsShim (upgrade key: %s)', upgradeKey);
+    }
+
+    if (endpoint.MozMobileConnectionsShim) {
+      endpoint.MozMobileConnectionsShim.init(
+        endpoint.navigator.mozMobileConnections,
+        upgradeKey ? new endpoint.BroadcastChannel(
+          `${endpoint.MozMobileConnectionsShim.name}-${upgradeKey}`
+        ) : endpoint
+      );
+
+      debug(
+        'Initialized MozMobileConnectionsShim (upgrade key: %s)',
+        upgradeKey
+      );
+    }
+  }
+
+  var shimHostPromise;
+  function getShimEndpoint(upgradeKey) {
+    debug('Request shim endpoint');
+
+    if (shimHostPromise) {
+      debug('Shim endpoint is ready');
+      return shimHostPromise;
+    }
+
+    if (upgradeKey) {
+      debug('Shim endpoint exists, but is not cached yet');
+      shimHostPromise = Promise.resolve(window.open('', upgradeKey));
+    } else {
+      debug('Shim endpoint does not exist yet');
+      shimHostPromise = new Promise((resolve) => {
+        var shimHost = window.open(
+          '/views/shared/shim_host.html', App.instanceId, 'alwaysLowered'
+        );
+
+        shimHost.addEventListener('load', function onLoad() {
+          shimHost.removeEventListener('load', onLoad);
+          resolve(shimHost);
+
+          debug('New shim endpoint is created and cached');
+        });
+      });
+    }
+
+    return shimHostPromise;
+  }
+
+  var ServiceManager = {
+    [priv.upgradeKey]: null,
+
+    initialize(serviceKeys) {
+      var upgradeKey = this[priv.upgradeKey] =
+        sessionStorage.getItem(UPGRADE_KEY) || null;
+
+      debug('Upgrade key: %s', upgradeKey);
+
+      if (!upgradeKey) {
+        initializeShims(exports, upgradeKey);
+      }
+
+      serviceKeys.forEach((serviceKey) => {
+        serviceInitializers.get(serviceKey).initialize(upgradeKey);
+      });
+    },
+
+    upgrade(serviceKeys) {
+      debug('Upgrade is requested');
+
+      if (this[priv.upgradeKey]) {
+        return Promise.resolve();
+      }
+
+      var upgradeKey = App.instanceId;
+
+      return getShimEndpoint().then((endpoint) => {
+        initializeShims(endpoint, upgradeKey);
```

Comment:
Not sure if it's a proper timing for shim init because we run upgrade when fully loaded. Some api like settings.setReadAheadThreadRetrieval might be needed before fully loaded, but we can definitely change the timing of calling API.